numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -342,9 +342,15 @@ array_implement_array_function(
     }
 
     /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
-     * in downstream libraries.
+     * in downstream libraries. If `like=` is specified but doesn't
+     * implement `__array_function__`, raise a `TypeError`.
      */
     if (kwargs != NULL && PyDict_Contains(kwargs, npy_ma_str_like)) {
+        PyObject *like_arg = PyDict_GetItem(kwargs, npy_ma_str_like);
+        if (like_arg && !get_array_function(like_arg)) {
+            return PyErr_Format(PyExc_TypeError,
+                "The `like` object must implement the `__array_function__` protocol");
```

Comment:
```suggestion
                "The `type(like)` must implement the `__array_function__` protocol");
```

Reply:
Good point to look at the message closer. Not sure like adding `type` it seems like unnecessary precision. I would suggest using the same jargon words we use in the documentation:

> must be an array-like that implements the ``__array_function__`` protocol, but is an instance of %S

(passing `Py_TYPE(obj)` for the format code. Although types print as `<class 'name'>` which is slightly annoying, and printing just the name was a bit more involved. So I am fine also with just not printing at all.)

Reply:
I'd say "expected" instead of "must be", but the rest is better!